### PR TITLE
Update for 1.0, drop Universe, add some tests, and env.total_reward 

### PR DESCRIPTION
**Pull request by JobJob**
Start of the changes to this and Reinforce that @tejank10 and I discussed at JuliaCon.

Edit: this requires ~some changes to Reinforce - changed to WIP for now~ https://github.com/JuliaML/Reinforce.jl/pull/23

Coming soon: speedups - Gym environments in Julia as fast as from Python.

### CODE REVIEW

**Comment by iblis17** [comment]
(cc myself)
I will find free time to review this.

**Comment by iblis17** [comment]
I want to drop 0.6 support, just go toward v0.7/1.0.
Update REQUIRE as well?

File: src/OpenAIGym.jl [hunk 1]
```diff
@@ -37,51 +38,17 @@ GymEnv(name) = gym(name)
 function Reinforce.reset!(env::GymEnv)
     env.state = env.pyenv[:reset]()
     env.reward = 0.0
+    env.total_reward = 0.0
     env.actions = actions(env, nothing)
     env.done = false
 end
 
-"A simple wrapper around the OpenAI gym environments to add to the Reinforce framework"
-type UniverseEnv <: AbstractGymEnv
```
[the remark this answers is not in the document]
Comment (JobJob) [reply src/OpenAIGym.jl]: I think Universe is basically dead, see here: https://github.com/openai/universe/issues/218#issuecomment-361711296 and the link there to this reddit comment by Ilya Sutskever here: https://www.reddit.com/r/MachineLearning/comments/6jjrk3/d_why_im_remaking_openai_universe/djfbk1b/

The Universe code being in the same file as the Gym code caught me out a couple of times when I edited the method for Universe instead of Gym, so if you really want to keep it I think it should be in a different file

[the remark this answers is not in the document]
Comment (iblis17) [reply src/OpenAIGym.jl]: Okay, deleting it is fine at this moment.
Is this the replacement? https://github.com/openai/retro

File: test/runtests.jl [hunk 2]
```diff
@@ -1,13 +1,136 @@
-using OpenAIGym
-using Base.Test
+using OpenAIGym, Compat.Test
```
Comment (iblis17) [inline test/runtests.jl]: I think `using Test` is fine now

**Comment by iblis17** [comment]
Could you enable this feature, so I can push some minor changes directly?
https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

**Comment by JobJob** [comment]
> Could you enable this feature, so I can push some minor changes directly?
https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

Yeah "Allow edits from maintainers." has always been checked

File: test/runtests.jl (hunk 2 above)
Comment (JobJob) [reply test/runtests.jl]: I'll leave you to do this and any other changes you like

File: src/OpenAIGym.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (JobJob) [reply src/OpenAIGym.jl]: No that's essentially just more gym environments (though would need a very small amount of setup code to get running in julia I think). The guy from that github comment has this: https://github.com/unixpickle/muniverse and said he's working at OpenAI, but nothing official has been released AFAIK

File: src/OpenAIGym.jl [hunk 3]
```diff
@@ -1,6 +1,3 @@
-
- __precompile__()
-
 module OpenAIGym
 
 using PyCall
```
Comment (JobJob) [inline src/OpenAIGym.jl]: this line wasn't added in this PR, but we need to add PyCall to Project.toml

Comment (iblis17) [reply src/OpenAIGym.jl]: I added PyCall here
https://github.com/JuliaML/OpenAIGym.jl/pull/10/commits/3058d7e4af3761c8fb3c0728b89e5f65882fe07d#diff-910a7b3ca0075d545dcec45cb7335ca9R7

**Comment by iblis17** [comment]
TODO
- [ ] We need to install gym in Travis to make the test pass.

File: src/OpenAIGym.jl [hunk 4]
```diff
@@ -19,14 +16,15 @@ const _py_envs = Dict{String,Any}()
...
+abstract type AbstractGymEnv <: AbstractEnvironment end
 
 "A simple wrapper around the OpenAI gym environments to add to the Reinforce framework"
-type GymEnv <: AbstractGymEnv
+mutable struct GymEnv <: AbstractGymEnv
     name::String
     pyenv  # the python "env" object
     state
     reward::Float64
+    total_reward::Float64
```
Comment (iblis17) [inline src/OpenAIGym.jl]: I think we should make them parametric in another PR.
e.g.
```julia
mutable struct GymEnv{T<:AbstracFloat}
  ...
  reward::T
  total_reward::T
  ...
```

Comment (JobJob) [reply src/OpenAIGym.jl]: I think the reward being Float64 is probably not such a big issue. The state type being parametric I have in #11

But yeah it probably doesn't hurt

Comment (iblis17) [reply src/OpenAIGym.jl]: :ok_hand:

**Comment by iblis17** [comment]
:zzz:

File: test/runtests.jl [hunk 5]
```diff
@@ -1,13 +1,137 @@
...
+    steps, t
+end
+
+"""
+Steps through an episode until it's `done`
+assumes env has been `reset!`
+"""
+function epstep(env::GymEnv)
+    steps = 0
+    while !env.done
```
Comment (iblis17) [inline test/runtests.jl]: I think `env.done` should be changed to `finished(env, s)`,
`env.done` is kind of internal variable, so user should not rely on it.

Comment (JobJob) [reply test/runtests.jl]: 👍 and nice, you got the tests working 🎉

**Comment by iblis17** [comment]
I read the code in Reinforce.jl again.
Found that there is a field `total_reward` available in `Reinforce.Epsode`,
so why we made similar stuffs in an environment?

**Comment by JobJob** [comment]
I don't really like the Episode abstraction, it was slow - that may be fixable, there were some type instabilities AFAIR, but I'm not sure it's really providing that much benefit. I'm open to being convinced though.

**Comment by JobJob** [comment]
Actually in general I'm of the view that this package should drop the dependency on Reinforce, and we create a separate package that wraps this package as a Reinforce environment, by just forwarding the relevant methods.

**Comment by JobJob** [comment]
The other thing about Episode, and having things in iterators in general, is that they sometimes make error messages slightly more obscure. Things are more transparent, and less magical, if you just have a loop.

Anyway I'm not wedded to the idea of total_reward but it's convenient IMO

**Comment by iblis17** [comment]
> I don't really like the Episode abstraction, it was slow - that may be fixable, there were some type instabilities AFAIR, but I'm not sure it's really providing that much benefit. I'm open to being convinced though.

> The other thing about Episode, and having things in iterators in general, is that they sometimes make error messages slightly more obscure. Things are more transparent, and less magical, if you just have a loop.

well, I want an abstraction. Acctually, I want an iteration protocol on episodic task.
I know `Reinforce.Episode` is slow and type unstable, but that's old unmaintained code.

Maybe we can draw a new blueprint about episodic iteration protocol and a concete implementation.
I want this implementation be generic and versitile. Everyone just adds features into it and other package gain new features.

I read the code of Episode again, the current features is barren:
1. it supports `maxn` to set the steps limitation.
2. nothing :( 

------
> Actually in general I'm of the view that this package should drop the dependency on Reinforce, and we create a separate package that wraps this package as a Reinforce environment, by just forwarding the relevant methods.

About the package design,
I think we can do this trade-off:
- Make OpenAIGym.jl plays the role of a Python library binding, so support Python's API is the primary goal.
- Add some integrations to Julia ecosystem is the sendary gaol, like Reinforce.jl / LearningStategy.jl

Let's put efforts on the primary  goal first.

**Comment by iblis17** [comment]
Given that we clearify the primary goal,
I can accept that you dropping the Reinforce stuffs at first,
even creating a GymEnv not subtyping the AbstractEnvrionment.

**Comment by JobJob** [comment]
Agreed re primary goal, this package should provide a **fast** way to access Gym environments from Julia via Python.

There is some interesting work going on for all the other stuff at https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl I haven't taken it all in yet, but the design looks quite nice.

**Comment by iblis17** [comment]
so... do you want to make some rework on this PR, or I can merge it now?

**Comment by JobJob** [comment]
Merge it! We can do other stuff later. 🙏

**Comment by iblis17** [comment]
what :interrobang:  I don't know this repo before...
https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl

**Comment by JobJob** [comment]
Haha, yeah I had the same reaction when I first stumbled upon it a month or so ago

File: src/OpenAIGym.jl [hunk 6]
```diff
@@ -140,49 +97,29 @@ function Reinforce.actions(env::AbstractGymEnv, s′)
...
-function test_env(name::String = "CartPole-v0")
-    env = gym(name)
-    for sars′ in Episode(env, RandomPolicy())
-        render(env)
-    end
-end
-
-
+global const pygym = PyNULL()
+global const pysoccer = PyNULL()
```
Comment (iblis17) [inline src/OpenAIGym.jl]: I think the keyword `global` is redundant.
I'm going to remove it.

File: src/OpenAIGym.jl [hunk 7]
```diff
@@ -37,51 +35,17 @@ GymEnv(name) = gym(name)
...
-    env.state = env.pyenv[:reset]()
-    env.reward = [0.0]
-    env.actions = actions(env, nothing)
-    env.done = [false]
-end
-
 function gym(name::AbstractString)
     env = if name in ("Soccer-v0", "SoccerEmptyGoal-v0")
-        @pyimport gym_soccer
+        Base.copy!(gym_soccer, pyimport("gym_soccer"))
```
Comment (iblis17) [inline src/OpenAIGym.jl]: correct this in #14

File: src/OpenAIGym.jl [hunk 8]
```diff
@@ -92,7 +56,8 @@ end
 
 # --------------------------------------------------------------
 
-render(env::AbstractGymEnv, args...) = env.pyenv[:render]()
+render(env::AbstractGymEnv, args...; kwargs...) =
+    pycall(env.pyenv[:render], PyAny; kwargs...)
```
Comment (iblis17) [inline src/OpenAIGym.jl]: I want to write this in `env.pyenv[:render](kwargs...)`,
 I guess there are no difference between them?
